app/src/main/java/com/group80/uoftinder/GenerateCompatibilityList.java

```diff
@@ -0,0 +1,113 @@
+import java.util.List;
+
+/**
+ * A use case that generates a list of compatible users for the current user
+ */
+public class GenerateCompatibilityList {
+    private User curUser;
+    private int curUserScore;
+    private List<String> compatibilityList;
+    private RecPresenterInterface recPresenterInterface;
+    private List<User> allUsers;
+    private UserScoreFacade usf;
+
+    /**
+     * Initialize the class by loading in all the users in the database, creating the
+     * UserScoreFacade instance, and getting the score of curUser
+     */
+    public GenerateCompatibilityList(RecPresenterInterface recPresenterInterface) {
+        List<User> allUsers = UserRealtimeDbFacade.getAllUsers(curUser.getUserType());
+        usf = new UserScoreFacade(curUser, new CreateAccountInteractor()); // may need to change 2nd param
+        User curUser = UserRealtimeDbFacade.getCurrentUser();
+        int curUserScore = curUser.getScore();
+        this.recPresenterInterface = recPresenterInterface;
+    }
+
+    /**
+     * Generate the compatibility list for curUser and update the User's compatibilityList attribute.
+     * Pass the compatibilityList to recPresenterInterface so that it can display it
+     */
+    public void generateCompatibilityList() {
+        compatibilityList = new ArrayList<>();
+        Map<String, Integer> compScores = calculateCompatibilityScores(allUsers);
+        for (int i = 0 ; i < users.size() ; i++) {
+            String maxKey = getMaxKey(compScores);
+            compatibilityList.add(maxKey);
+            compScores.remove(maxKey);
+        }
+//        curUser.setCompatibilityList(compatibilityList);
+//        recPresenterInterface.showUsers(compatibilityList);
+    }
+
+    /**
+     * Return the key that corresponds to the max value in userScore
+     * @param userScore: a map that maps the user ID to their compatibility score with the current
+     *                 user
+     * @return the key that corresponds to the max value in userScore
+     */
+    private String getMaxKey(Map<String, Integer> userScore) {
+        // TODO: change this to get key of max value
+        int maxValue = 0;
+        String maxKey = "";
+        for (String key : userScore.keySet()) {
+            Integer currentScore = userScore.get(key);
+            if (currentScore != null && currentScore > maxValue) {
+//                if (currentScore > maxValue) {
+                maxValue = currentScore;
+                maxKey = key;
+//                }
+            }
+
+        }
+        return maxKey;
+    }
+
+    /**
+     * Return a map of user ID to the user's compatibility score with the current user
+     * @param users: a list of all User in the database
+     * @return a map of user ID to the user's compatibility score with the current user
+     */
+    private Map<String, Integer> calculateCompatibilityScores(List<User> users) {
+        Map<String, Integer> compScores = new Hashtable<>();
```

Comment:
Just a question - was there any rationale behind choosing Hashtable over HashMap here?

Reply:
There's no specific reason. Do you think it would be better to use a HashMap?

Reply:
Ah I see. Did some quick research - "HashMap is not synchronized, therefore it's faster and uses less memory than Hashtable." If `compScores ` is just a local variable and not a resource shared by other sources/threads, I think HashMap might be the more efficient option!